### ISSUE TITLE
GRPOTrainer generates garbage output for models requiring BOS tokens due to add_special_tokens=False

### ISSUE THREAD

**Issue by CuriousG102** [huggingface/trl#3520]
### Reproduction

# GRPOTrainer generates garbage output for models requiring BOS tokens due to add_special_tokens=False

## Description

GRPOTrainer calls the tokenizer with `add_special_tokens=False` in its `_prepare_inputs()` method (line 525), which removes special tokens like BOS (`<bos>`) that some models (e.g., Gemma family) require for proper text generation. This causes the model to generate garbage, repetitive patterns, or nonsensical text during training, even though the same model generates correctly outside the trainer.

## Minimal Reproduction

```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM
from datasets import Dataset
from trl import GRPOTrainer, GRPOConfig
from peft import LoraConfig, get_peft_model

# Load a model that requires BOS tokens
model_name = "google/gemma-3-1b-it"
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Use appropriate dtype for device
if torch.backends.mps.is_available():
    dtype = torch.float32  # MPS doesn't support float16 well
    device = "mps"
elif torch.cuda.is_available():
    dtype = torch.float16
    device = "cuda"
else:
    dtype = torch.float32
    device = "cpu"

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    torch_dtype=dtype,
    device_map=device,
)

# Apply LoRA
peft_config = LoraConfig(r=8, lora_alpha=16, target_modules=["q_proj", "v_proj"])
model = get_peft_model(model, peft_config)

# Create dataset
dataset = Dataset.from_dict({
    "prompt": ["What is the capital of France?", "What is 2 + 2?"]
})

# Capture generated texts
generated_texts = []
def capture_reward(prompts, completions, **kwargs):
    generated_texts.extend(completions)
    return [0.0] * len(prompts)

# Train with GRPOTrainer
config = GRPOConfig(
    output_dir="./test_output",
    per_device_train_batch_size=2,
    num_generations=2,
    max_steps=1,
    report_to=[],
)

trainer = GRPOTrainer(
    model=model,
    args=config,
    train_dataset=dataset,
    processing_class=tokenizer,
    reward_funcs=[capture_reward],
)

trainer.train()

# Check generated texts
for i, text in enumerate(generated_texts):
    print(f"Generated [{i}]: {text}")
    # Will show garbage/repetitive output
```

## Expected Behavior

The model should generate coherent responses like "The capital of France is Paris" during training, just as it does when used directly.

## Actual Behavior

The model generates garbage text, repetitive patterns, or nonsensical output because the BOS token is missing.

Example outputs:
- Repetitive dots: `"..........................."`
- Nonsense: `"    \n\n\n\n\n    "`
- Broken text: `"The the the the the"`

## Root Cause

In `GRPOTrainer._prepare_inputs()` (around line 525):

```python
prompt_inputs = self.processing_class(
    prompts_text, return_tensors="pt", padding=True, padding_side="left", add_special_tokens=False
)
```

The `add_special_tokens=False` parameter removes the BOS token that models like Gemma require.

## Proposed Fix

Either:
1. Change to `add_special_tokens=True`
2. Make it configurable via GRPOConfig
3. Auto-detect based on model requirements

## Workaround

Users can wrap their tokenizer to force `add_special_tokens=True`:

```python
class TokenizerWrapper:
    def __init__(self, tokenizer):
        self._wrapped = tokenizer
    
    def __call__(self, *args, add_special_tokens=True, **kwargs):
        if not add_special_tokens:
            add_special_tokens = True
        return self._wrapped(*args, add_special_tokens=add_special_tokens, **kwargs)
    
    def __getattr__(self, name):
        return getattr(self._wrapped, name)

# Use wrapped tokenizer
wrapped_tokenizer = TokenizerWrapper(tokenizer)
trainer = GRPOTrainer(
    model=model,
    processing_class=wrapped_tokenizer,
    # ... other args
)
```

## Environment

`trl env` output:
```
% trl env
INFO 06-01 16:24:39 [__init__.py:239] Automatically detected platform cpu.

Copy-paste the following information when reporting an issue:

- Platform: macOS-15.5-arm64-arm-64bit
- Python version: 3.10.16
- PyTorch version: 2.6.0
- CUDA device(s): not available
- Transformers version: 4.51.0
- Accelerate version: 1.6.0
- Accelerate config: not found
- Datasets version: 3.5.0
- HF Hub version: 0.30.1
- TRL version: 0.15.2
- bitsandbytes version: not installed
- DeepSpeed version: not installed
- Diffusers version: 0.32.2
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.70.0
- PEFT version: 0.15.1
```

## Additional Context

This issue affects any model that relies on special tokens for proper generation, particularly:
- Google Gemma-3 models (gemma-3-1b-it, gemma-3-4b-it)
- Potentially other models with similar tokenization requirements

The issue is particularly problematic because:
1. The same model works fine outside GRPOTrainer
2. There's no clear error message - just poor generation quality
3. It's not obvious that `add_special_tokens=False` is the cause

### System Info

```
% trl env
INFO 06-01 16:24:39 [__init__.py:239] Automatically detected platform cpu.

Copy-paste the following information when reporting an issue:

- Platform: macOS-15.5-arm64-arm-64bit
- Python version: 3.10.16
- PyTorch version: 2.6.0
- CUDA device(s): not available
- Transformers version: 4.51.0
- Accelerate version: 1.6.0
- Accelerate config: not found
- Datasets version: 3.5.0
- HF Hub version: 0.30.1
- TRL version: 0.15.2
- bitsandbytes version: not installed
- DeepSpeed version: not installed
- Diffusers version: 0.32.2
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.70.0
- PEFT version: 0.15.1
```

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by behroozazarkhalili** [comment]
I think this one may be resolved by refactoring since it was filed. Checking against current `main`:

The generation path no longer tokenizes prompts with `add_special_tokens=False`. `_generate` now builds prompt token ids through `self.processing_class.apply_chat_template(conversation=prompts, add_generation_prompt=True, tokenize=True, ...)` (`grpo_trainer.py:1510`), so BOS and other special tokens are emitted by the chat template rather than stripped. The `_prepare_inputs` line 525 call this issue pointed at (with `add_special_tokens=False`) is gone.

The only remaining `add_special_tokens=False` in the file is at `grpo_trainer.py:1427`, and it is in the reward-model scoring path (tokenizing `prompt + completion` to feed a reward model), not the generation path, so it would not cause the garbage generation described here.

One caveat on the original repro: it used a raw-string `prompt` dataset (`Dataset.from_dict({"prompt": [...]})`) rather than a chat/conversational one. With a conversational dataset the template supplies BOS; the raw-string path is a different code route, so it would be worth confirming the Gemma repro still reproduces on current `main` before closing.

@qgallouedec does this match your read? If the generation-path change already covers it, this looks closeable.

**Comment by albertvillanova** [comment]
Hey @CuriousG102, sorry for the slow reply here — we are catching up on the backlog. 

Good news: this is already fixed on current `main`. The `_prepare_inputs` tokenizer call you pointed at (line 525 in 0.15.2, with `add_special_tokens=False`) is gone: conversational prompts are now tokenized via `apply_chat_template(..., tokenize=True, add_generation_prompt=True)` so BOS comes from the template, and raw-string `prompt` datasets like the one in your repro go through `self.processing_class(text=prompts)` with the default `add_special_tokens=True` — see https://github.com/huggingface/trl/blob/1925183c91852c30088ed8ad53181e7b7147489e/trl/trainer/grpo_trainer.py#L1636. I confirmed with the Gemma-3 tokenizer that this path now leads with the `<bos>` id where the old one dropped it, so generation should no longer collapse into repetitive/garbage text. The only remaining `add_special_tokens=False` is in the reward-model scoring path (https://github.com/huggingface/trl/blob/1925183c91852c30088ed8ad53181e7b7147489e/trl/trainer/grpo_trainer.py#L1505-L1507), which does not affect generation. 

Closing as fixed (thanks @behroozazarkhalili for the earlier triage); please re-open if you still see garbage on a current TRL version.
